mlflow/genai/judges/instructions_judge/__init__.py

```diff
@@ -0,0 +1,271 @@
+import json
+from typing import Any
+
+from pydantic import PrivateAttr
+
+from mlflow.entities.model_registry.prompt_version import PromptVersion
+from mlflow.entities.trace import Trace
+from mlflow.exceptions import MlflowException
+from mlflow.genai.judges.base import Judge
+from mlflow.genai.judges.utils import format_prompt, get_default_model, invoke_judge_model
+from mlflow.genai.scorers.base import ScorerKind
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.utils.annotations import experimental
+
+
+@experimental(version="3.4.0")
+class InstructionsJudge(Judge):
+    """
+    A judge that evaluates traces based on user-provided instructions.
+
+    This judge uses natural language instructions to guide evaluation,
+    making it flexible for various assessment criteria.
+    """
+
+    _TEMPLATE_VARIABLE_INPUTS = "inputs"
+    _TEMPLATE_VARIABLE_OUTPUTS = "outputs"
+    _TEMPLATE_VARIABLE_TRACE = "trace"
+    _TEMPLATE_VARIABLE_EXPECTATIONS = "expectations"
+    _RESERVED_INSTRUCTION_TEMPLATE_VARIABLES = [
+        _TEMPLATE_VARIABLE_INPUTS,
+        _TEMPLATE_VARIABLE_OUTPUTS,
+        _TEMPLATE_VARIABLE_TRACE,
+        _TEMPLATE_VARIABLE_EXPECTATIONS,
+    ]
+
+    _instructions: str = PrivateAttr()
+    _model: str = PrivateAttr()
+    _instructions_prompt: PromptVersion = PrivateAttr()
+    _custom_template_variables: set[str] = PrivateAttr()
+
+    def __init__(self, name: str, instructions: str, model: str | None = None, **kwargs):
+        """
+        Initialize the InstructionsJudge.
+
+        Args:
+            name: The name of the judge
+            instructions: Natural language instructions for evaluation
+            model: The model identifier to use for evaluation (e.g., "openai/gpt-4")
+            kwargs: Additional configuration parameters
+        """
+        super().__init__(name=name, **kwargs)
+
+        if not name or not isinstance(name, str):
+            raise MlflowException(
+                "name must be a non-empty string", error_code=INVALID_PARAMETER_VALUE
+            )
+        if not instructions or not isinstance(instructions, str):
+            raise MlflowException(
+                "instructions must be a non-empty string", error_code=INVALID_PARAMETER_VALUE
+            )
+
+        self._instructions = instructions
+        self._model = model or get_default_model()
+
+        self._instructions_prompt = PromptVersion(
```

Comment:
Can we add an inline comment explaining why we construct a dummy prompt version here?